### PR TITLE
[BugFix] Fix several struct type bugs

### PR DESCRIPTION
**Pull request by Smith-Cruise**
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes:

1. Throw errors when user use 'struct_field' instead of `struct_field`: #13449
2. Fix subfield materialized bug: #13484 #13451 #13592 
3. Struct subfield name should be case-insensitive, the same as Trino and Hive: #13455
4. Fix column name NullPointer bug: #13589

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
## Please check the version labels which this pr will be auto backported to target branch and click the checkbox:

- [ ] check done.

**Comment by Smith-Cruise** [comment]
run pr-title-check

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 31 / 32 (96.88%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/StructType.java|3|4|75.00%|[58]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/SubfieldOperator.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/SubfieldExpr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/StructField.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/Field.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|13|13|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|2|2|100.00%|[]|

**Review by Youngwb — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 33 / 34 (97.06%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/StructType.java|4|5|80.00%|[57]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/SubfieldOperator.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/SubfieldExpr.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|13|13|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|2|2|100.00%|[]|

**Review by Youngwb — DISMISSED** [review]

**Review by dirtysalt — DISMISSED** [review]

**Comment by Smith-Cruise** [comment]
run starrocks_fe_unittest

File: fe/fe-core/src/main/java/com/starrocks/analysis/SlotRef.java [hunk 1]
```diff
@@ -107,11 +107,12 @@ public SlotRef(QualifiedName qualifiedName) {
             this.col = parts.get(3);
             this.label = parts.get(3);
         } else {
+            // If parts.size() > 4, it must refer to a struct subfield name, so we set SlotRef's TableName null value,
+            // set col, label a qualified name here[Of course it's a wrong value].
+            // Correct value will be parsed in Analyzer according context.
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/analysis/SlotRef.java:112]: if it's wrong, why not just give it empty string.

Comment (Smith-Cruise) [reply fe/fe-core/src/main/java/com/starrocks/analysis/SlotRef.java:112]: We need to return the error msg to the user if you set col = "", the error msg will be empty too.

File: fe/fe-core/src/main/java/com/starrocks/catalog/StructType.java [hunk 2]
```diff
@@ -52,12 +52,14 @@ public StructType(ArrayList<StructField> structFields) {
         Preconditions.checkArgument(structFields.size() > 0);
         this.fields = new ArrayList<>();
         for (StructField field : structFields) {
-            if (fieldMap.containsKey(field.getName())) {
-                LOG.warn(String.format("Contains the same struct subfield name: %s, ignore it", field.getName()));
+            String fieldName = field.getName();
+            if (fieldMap.containsKey(fieldName)) {
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/catalog/StructType.java:56]: should call `toLowerCase`

**Review by DorianZheng — CHANGES_REQUESTED** [review]
left comments

**Review by Youngwb — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 35 / 36 (97.22%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/StructType.java|4|5|80.00%|[57]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/SubfieldOperator.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/SubfieldExpr.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/SlotRef.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|13|13|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|2|2|100.00%|[]|

**Review by DorianZheng — APPROVED** [review]
LGTM

**Review by Youngwb — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by Smith-Cruise** [comment]
run starrocks_admit_test

**Comment by Smith-Cruise** [comment]
@mergify backport branch-2.5

**Pull request merged**